monitoredresources/monitored_resources.go

```diff
@@ -0,0 +1,169 @@
+		GCELabelZone:       gce.Zone,
+	}
+	return labels
+}
+
+type AWSEC2Instance struct {
+	AWSAccount string
+	InstanceID string
+	Region     string
+}
+
+func (aws *AWSEC2Instance) GetType() string {
+	return "aws_ec2_instance"
+}
+
+func (aws *AWSEC2Instance) GetLabels() map[string]string {
+	labels := map[string]string{
+		AWSEC2LabelAwsAccount: aws.AWSAccount,
+		AWSEC2LabelInstanceID: aws.InstanceID,
+		AWSEC2LabelRegion:     aws.Region,
+	}
+	return labels
+}
+
+var mr MonitoredResInterface
+
+// Autodetect auto detects monitored resources based on
+// the environment where the application is running.
+// It supports detection of following resource types
+// 1. gke_container:
+// 2. gce_instance:
+// 3. aws_ec2_instance:
+//
+// Returns MonitoredResInterface which implements getLabels() and getType()
+// For resource definition go to https://cloud.google.com/monitoring/api/resources
+func Autodetect() MonitoredResInterface {
+	detectOnce()
+	return mr
+}
+
+// createAWSEC2InstanceMonitoredResource creates a aws_ec2_instance monitored resource
+func createAWSEC2InstanceMonitoredResource() *AWSEC2Instance {
+	awsIdentityDoc := getAWSIdentityDocument()
+	aws_instance := AWSEC2Instance{
+		AWSAccount: awsIdentityDoc.AccountId,
+		InstanceID: awsIdentityDoc.InstanceId,
+		Region:     fmt.Sprintf("aws:%s", awsIdentityDoc.Region),
+	}
+	return &aws_instance
+}
+
+// createGCEInstanceMonitoredResource creates a gce_instance monitored resource
+func createGCEInstanceMonitoredResource() *GCEInstance {
+	gce_instance := GCEInstance{
+		ProjectID:  getGCPMetadataConfig(gcpProjectID),
+		InstanceID: getGCPMetadataConfig(gcpInstanceID),
+		Zone:       getGCPMetadataConfig(gcpZone),
+	}
+	return &gce_instance
+}
+
+// createGKEContainerMonitoredResource creates a gke_container monitored resource
+func createGKEContainerMonitoredResource() *GKEContainer {
+	gke_container := GKEContainer{
+
+		ProjectID:     getGCPMetadataConfig(gcpProjectID),
+		InstanceID:    getGCPMetadataConfig(gcpInstanceID),
+		ClusterName:   getGCPMetadataConfig(gcpClusterName),
+		ContainerName: getGCPMetadataConfig(gcpContainerNameEnvVar),
+		NamespaceID:   getGCPMetadataConfig(gcpNamespaceEnvVar),
+		PodId:         getGCPMetadataConfig(gcpPodIDEnvVar),
+		Zone:          getGCPMetadataConfig(gcpZone),
+	}
+	return &gke_container
+}
+
+var once sync.Once
+
+// detectResourceType determines the resource type.
+func detectResourceType() {
+	if os.Getenv("KUBERNETES_SERVICE_HOST") != "" {
+		mr = createGKEContainerMonitoredResource()
+	} else if getGCPMetadataConfig(gcpInstanceID) != "" {
+		mr = createGCEInstanceMonitoredResource()
+	} else if isRunningOnAwsEc2() {
+		mr = createAWSEC2InstanceMonitoredResource()
+	} else {
+		mr = nil
+	}
+}
+
+// detectOnce runs only once to detect the resource type.
+// It first attempts to retrieve AWS Identity Doc and GCP metadata.
+// It then determines the resource type
+func detectOnce() {
+	once.Do(func() {
+		retrieveAWSIdentityDocument()
```

Comment:
Can we do these in parallel and return the first once that succeeds?

Reply:
it doesn't take that long to detect. so parallel run is not required. Even parallel run is added the evaluation has to happen in specific order so it cannot be the first one to complete.